odl/discr/lp_discr.py

```diff
@@ -686,65 +684,430 @@ def __setitem__(self, indices, values):
+
+        if method == '__call__':
+            if ufunc.nout == 1:
+                kwargs['out'] = (out,)
+                res_tens = self.tensor.__array_ufunc__(
+                    ufunc, '__call__', *input_tensors, **kwargs)
+
+                if out is None:
+                    # Wrap result tensor in appropriate DiscreteLp space.
+                    # Make new function space based on result dtype,
+                    # keep everything else, and get `dspace` from the result
+                    # tensor.
+                    out_dtype = (res_tens.dtype, self.space.uspace.out_shape)
+                    fspace = FunctionSpace(self.space.uspace.domain,
+                                           out_dtype)
+                    res_space = DiscreteLp(
+                        fspace, self.space.partition,
+                        res_tens.space, self.space.interp_byaxis,
+                        axis_labels=self.space.axis_labels)
+                    result = res_space.element(res_tens)
+                else:
+                    result = out_tuple[0]
+
+                return result
+
+            elif ufunc.nout == 2:
+                kwargs['out'] = (out1, out2)
+                res1_tens, res2_tens = self.tensor.__array_ufunc__(
+                    ufunc, '__call__', *input_tensors, **kwargs)
+
+                if out1 is None:
+                    # Wrap as for nout = 1
+                    out_dtype = (res1_tens.dtype, self.space.uspace.out_shape)
+                    fspace = FunctionSpace(self.space.uspace.domain,
+                                           out_dtype)
+                    res_space = DiscreteLp(
+                        fspace, self.space.partition,
+                        res1_tens.space, self.space.interp_byaxis,
+                        axis_labels=self.space.axis_labels)
+                    result1 = res_space.element(res1_tens)
+                else:
+                    result1 = out_tuple[0]
+
+                if out2 is None:
+                    # Wrap as for nout = 1
+                    out_dtype = (res2_tens.dtype, self.space.uspace.out_shape)
+                    fspace = FunctionSpace(self.space.uspace.domain,
+                                           out_dtype)
+                    res_space = DiscreteLp(
+                        fspace, self.space.partition,
+                        res2_tens.space, self.space.interp_byaxis,
+                        axis_labels=self.space.axis_labels)
+                    result2 = res_space.element(res2_tens)
+                else:
+                    result2 = out_tuple[1]
 
-        There is also support for various reductions (sum, prod, min, max)
+                return result1, result2
 
-        >>> x.ufuncs.sum()
-        -1.0
+            else:
+                raise NotImplementedError('nout = {} not supported'
+                                          ''.format(ufunc.nout))
+
+        elif method == 'reduce' and keepdims:
```

Comment:
Well, your example is for a `TensorSpace`, where this is meaningful and allowed. But what would you do with a `DiscreteLp` where you would have to collapse an axis to a single point? I don't see a meaningful way to do this.

Reply:
Part of the answer is that the result in your example is wrong, it should be `rn(1).element([3.0])`. The `keepdims=True` option means "keep axes of size 1 instead of removing them". It does **not** mean "broadcast the result to the original shape".

Reply:
It seems you are right, agree with this then.

Reply:
Just for the record: for the same reason, `outer` requires two `DiscreteLpElement` objects, since otherwise we don't know how to infer the new space. A similar reason applies to `reduceat`.